### PR TITLE
validation: static_assert to ensure width in unit class

### PR DESCRIPTION
**Pull request by shargon**
Because operators are implemented in the base class, and being inherited by uint256 and unit512, if a copy or read is made between these types, random memory would be being accessed.

### CODE REVIEW

**Comment by tryphe** [comment]
Why skip the other functions, like `operator =`? It seems like the same rationale should apply. Just curious.

**Comment by shargon** [comment]
@tryphe Thanks, was forgotten

**Comment by laanwj** [comment]
ACK, looks good to me, these sanity asserts can't hurt.

Please squash your commits.

**Comment by laanwj** [comment]
This gives me the following compile error (clang linux)
```
/…/bitcoin/src/arith_uint256.h:51:32: error: static_assert expression is not an integral constant expression
        static_assert(WIDTH == b.WIDTH, "Template parameter WIDTH must be equal.");
                      ~~~~~~~~~^~~~~~~
/home/orion/projects/bitcoin/bitcoin/src/arith_uint256.h:268:7: note: in instantiation of member function 'base_uint<256>::operator=' requested here
class arith_uint256 : public base_uint<256> {
      ^
1 error generated
```

**Comment by practicalswift** [comment]
Concept ACK: compile time checking cannot hurt

**Comment by shellenrek** [comment]
unsubscribe

发送自 Windows 10 版邮件<https://go.microsoft.com/fwlink/?LinkId=550986>应用

________________________________
发件人: practicalswift <notifications@github.com>
发送时间: Monday, October 28, 2019 10:56:04 PM
收件人: bitcoin/bitcoin <bitcoin@noreply.github.com>
抄送: Subscribed <subscribed@noreply.github.com>
主题: Re: [bitcoin/bitcoin] validation: static_assert to ensure width in unit class (#16696)


Concept ACK: compile time checking cannot hurt

―

**Comment by MarcoFalke** [comment]
@shargon Are you still working on this?

**Comment by shargon** [comment]
I am on vacations, please give me a week.

**Comment by shargon** [comment]
Could you summarize the expected changes and confirm that it could be merged in order to avoid the waste of time as #16310

**Comment by MarcoFalke** [comment]
Adding `static_asserts` to document assumptions is nice, but they must have no false positives.

**Comment by fanquake** [comment]
Let me know if you want this reopened.
